File: sdks/python/tests/e2e_live.py
"""Live end-to-end test: Python SDK -> beaterd OTLP -> read API.

Emits a real agent trace through the ergonomic SDK, then queries the running
beaterd HTTP API to prove the trace landed and the span kinds/attributes
normalized correctly. Driven by scripts/e2e-sdk-live.sh against a live server.

Exit code 0 = pass.
"""

import json
import os
import sys
import time
import urllib.request

from opentelemetry import trace as otel_trace

import beater
from beater import SpanKind

BASE_URL = os.environ["BEATER_BASE_URL"].rstrip("/")
TENANT = os.environ.get("BEATER_TENANT_ID", "demo")


def get_json(path: str):
    with urllib.request.urlopen(f"{BASE_URL}{path}", timeout=10) as resp:
        return json.loads(resp.read().decode())


def emit_trace() -> str:
    beater.init(service_name="beater-e2e", release_id="e2e")
    captured = {}

    @beater.observe(kind=SpanKind.AGENT_RUN, name="e2e-run")
    def run():
        captured["trace_id"] = format(
            otel_trace.get_current_span().get_span_context().trace_id, "032x"
        )
        plan()
        call_model()

    @beater.observe(kind=SpanKind.AGENT_PLAN, name="e2e-plan")
    def plan():
        beater.set_output("plan: look up policy")

    @beater.observe(kind=SpanKind.LLM_CALL, name="e2e-llm")
    def call_model():
        span = otel_trace.get_current_span()
        span.set_attribute("llm.provider", "openai")
        span.set_attribute("llm.model_name", "gpt-e2e")
        span.set_attribute("llm.token_count.prompt", 12)
        span.set_attribute("llm.token_count.completion", 5)
        beater.set_output("escalate")

    run()
    beater.flush()
    return captured["trace_id"]


def main() -> int:
    trace_id = emit_trace()
    print(f"emitted trace_id={trace_id}")

    # Poll the read API until the trace is queryable (ingest is async/buffered).
    deadline = time.time() + 30
    trace = None
    while time.time() < deadline:
        try:
            trace = get_json(f"/v1/traces/{TENANT}/{trace_id}")
            if trace.get("spans"):
                break
        except Exception:
            pass
        time.sleep(1)

    if not trace or not trace.get("spans"):
        print("FAIL: trace never became queryable", file=sys.stderr)
        return 1

    kinds = {s["kind"] for s in trace["spans"]}
    print(f"landed spans: {len(trace['spans'])} kinds={sorted(kinds)}")

    expected = {"agent.run", "agent.plan", "llm.call"}
    missing = expected - kinds
    if missing:
        print(f"FAIL: missing span kinds {missing}", file=sys.stderr)
        return 1

    llm = next((s for s in trace["spans"] if s["kind"] == "llm.call"), None)
    if not llm or not llm.get("model") or not llm.get("tokens"):
        print(f"FAIL: llm span missing model/tokens: {llm}", file=sys.stderr)
        return 1

    print("PASS: SDK -> beaterd -> read API round-trip verified")
    print(f"  llm model={llm['model']} tokens={llm['tokens']}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdks/typescript/tests/e2e_live.ts
// Live E2E: TS SDK -> paletted OTLP/HTTP -> read API. Run by scripts/e2e-sdk-live-ts.sh
// against a running paletted. Exit 0 = pass.

import { trace } from "@opentelemetry/api";

import * as palette from "../src/index";
import { SpanKind } from "../src/semconv";

const BASE_URL = (process.env.PALETTE_BASE_URL ?? "http://127.0.0.1:18080").replace(/\/+$/, "");
const TENANT = process.env.PALETTE_TENANT_ID ?? "demo";

async function getJson(path: string): Promise<any> {
  const res = await fetch(`${BASE_URL}${path}`);
  if (!res.ok) throw new Error(`${path} -> ${res.status}`);
  return res.json();
}

async function main(): Promise<number> {
  palette.init({ serviceName: "palette-e2e-ts", releaseId: "e2e-ts" });

  let traceId = "";
  const run = palette.observe(
    async () => {
      traceId = trace.getActiveSpan()!.spanContext().traceId;
      const plan = palette.observe(async () => palette.setOutput("plan"), {
        kind: SpanKind.AGENT_PLAN,
        name: "ts-plan",
      });
      await plan();
      const llm = palette.observe(
        async () => {
          const s = trace.getActiveSpan()!;
          s.setAttribute("llm.provider", "anthropic");
          s.setAttribute("llm.model_name", "claude-e2e");
          s.setAttribute("llm.token_count.prompt", 9);
          s.setAttribute("llm.token_count.completion", 4);
          palette.setOutput("done");
        },
        { kind: SpanKind.LLM_CALL, name: "ts-llm" },
      );
      await llm();
    },
    { kind: SpanKind.AGENT_RUN, name: "ts-run" },
  );

  await run();
  await palette.flush();
  console.log(`emitted trace_id=${traceId}`);

  const deadline = Date.now() + 30_000;
  let traceView: any;
  while (Date.now() < deadline) {
    try {
      traceView = await getJson(`/v1/traces/${TENANT}/${traceId}`);
      if (traceView?.spans?.length) break;
    } catch {
      /* retry */
    }
    await new Promise((r) => setTimeout(r, 1000));
  }

  if (!traceView?.spans?.length) {
    console.error("FAIL: trace never became queryable");
    return 1;
  }
  const kinds = new Set<string>(traceView.spans.map((s: any) => s.kind));
  console.log(`landed spans: ${traceView.spans.length} kinds=${[...kinds].sort()}`);
  for (const k of ["agent.run", "agent.plan", "llm.call"]) {
    if (!kinds.has(k)) {
      console.error(`FAIL: missing kind ${k}`);
      return 1;
    }
  }
  console.log("PASS: TS SDK -> paletted -> read API round-trip verified");
  return 0;
}

process.on("unhandledRejection", (reason) => {
  console.error("background rejection (non-fatal export retry?):", reason);
});

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    console.error("FATAL in main:", err);
    process.exit(1);
  });
